### Importing the libraries and Dataset

In [78]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import mean_absolute_error,mean_squared_error

In [69]:
def get_accuracy(y_pred,y_test):
    correctly_classified=0
    for count in range(np.size(y_pred)) :
        if y_test[count]==y_pred[count]:
           correctly_classified=correctly_classified+1
    acc=(correctly_classified/len(y_pred))*100
    print("Accuracy on test set is {:.2f}".format(acc))
    return acc 

In [70]:
df=load_iris(return_X_y=True,as_frame=True)[0]
df=df*10 #transfomring into mm
df=df.rename({"sepal length (cm)" : "sepal length (mm)", "sepal width (cm)" : "sepal width (mm)","petal length (cm)" : "petal length (mm)","petal width (cm)" : "petal width (mm)"},axis='columns')
df=df.astype(int)
df

sepal length (mm)  sepal width (mm)  petal length (mm)  petal width (mm)
0                   51                35                 14                 2
1                   49                30                 14                 2
2                   47                32                 13                 2
3                   46                31                 15                 2
4                   50                36                 14                 2
..                 ...               ...                ...               ...
145                 67                30                 52                23
146                 63                25                 50                19
147                 65                30                 52                20
148                 62                34                 54                23
149                 59                30                 51                18

[150 rows x 4 columns]

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   sepal length (mm)  150 non-null    int32
 1   sepal width (mm)   150 non-null    int32
 2   petal length (mm)  150 non-null    int32
 3   petal width (mm)   150 non-null    int32
dtypes: int32(4)
memory usage: 2.5 KB


### 1) Training the model

In [72]:
x=df.iloc[:,:-1].values
y=df.iloc[:,-1:].values

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)

In [76]:
sk_model = LogisticRegression()

sk_model.fit(x_train,y_train)

y_pred=sk_model.predict(x_test)

C:\Users\Crash\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\utils\validation.py:993: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\Crash\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [77]:
acc=get_accuracy(y_pred,y_test)

Accuracy on test set is 23.33


### 2) Classification mertics

MAE :

In [80]:
mean_absolute_error(y_test,y_pred)

1.9666666666666666

MSE :

In [81]:
mean_squared_error(y_test,y_pred)

7.433333333333334

RMSE :

In [82]:
mean_squared_error(y_test,y_pred)**(1/2)

2.7264140062238043

### 3) Commenting the results :

First of all we got an accuracy of 23.33% which is bad , but we should keep in mind that our accuracy function calculates how many times our model predicted the EXACT value, so we should check how far off our model is from the actual values, and we do that using our Classification metrics.
We can see that we got MAE,MSE and RMSE values that are considerably big, which means our predictions are very far off the actual values , thus our model is a bad one and needs to be replaced or modified (which is unlikely since it's really far off).